UnityProject/Assets/Scripts/Tilemaps/Behaviours/Layers/MetaTileMap.cs

```diff
@@ -1,587 +1,1316 @@
 			{
-				Layers[layer.LayerType].SetPreviewTile(position, LayerTile.EmptyTile, Matrix4x4.identity);
+				Layer layer = LayersValues[i];
+				if (layer.LayerType < tile.LayerType)
+				{
+					Layers[layer.LayerType].SetPreviewTile(position, LayerTile.EmptyTile, Matrix4x4.identity);
+				}
 			}
+
+			if (!Layers.ContainsKey(tile.LayerType))
+			{
+				Logger.LogErrorFormat($"LAYER TYPE: {0} not found!", Category.TileMaps, tile.LayerType);
+				return;
+			}
+
+			Layers[tile.LayerType].SetPreviewTile(position, tile, transformMatrix);
 		}
 
-		if (!Layers.ContainsKey(tile.LayerType))
+		public void ClearPreview()
 		{
-			Logger.LogErrorFormat($"LAYER TYPE: {0} not found!", Category.TileMaps, tile.LayerType);
-			return;
+			for (var i = 0; i < LayersValues.Length; i++)
+			{
+				LayersValues[i].ClearPreview();
+			}
 		}
+#endif
 
-		Layers[tile.LayerType].SetPreviewTile(position, tile, transformMatrix);
-	}
+		#region Physics
 
-	public void ClearPreview()
-	{
-		for (var i = 0; i < LayersValues.Length; i++)
+		//Gets the first hit
+		public MatrixManager.CustomPhysicsHit? Raycast(
+			Vector2 origin,
+			Vector2 direction,
+			float distance,
+			LayerTypeSelection layerMask, Vector2? To = null)
 		{
-			LayersValues[i].ClearPreview();
+			if (To == null)
+			{
+				To = direction.normalized * distance;
+			}
+
+			if (direction.x == 0 && direction.y == 0)
+			{
+				direction = (To.Value - origin).normalized;
+			}
+
+			// var Beginning = (new Vector3((float) origin.x, (float) origin.y, 0).ToWorld(PresentMatrix));
+			// Debug.DrawLine(Beginning + (Vector3.right * 0.09f), Beginning + (Vector3.left * 0.09f), Color.yellow, 30);
+			// Debug.DrawLine(Beginning + (Vector3.up * 0.09f), Beginning + (Vector3.down * 0.09f), Color.yellow, 30);
+
+			// var end = (new Vector3((float) To.Value.x, (float) To.Value.y, 0).ToWorld(PresentMatrix));
+			// Debug.DrawLine(end + (Vector3.right * 0.09f), end + (Vector3.left * 0.09f), Color.red, 30);
+			// Debug.DrawLine(end + (Vector3.up * 0.09f), end + (Vector3.down * 0.09f), Color.red, 30);
+
+			// Debug.DrawLine(Beginning, end, Color.magenta, 30);
+
+
+			Vector2 Relativetarget = To.Value - origin;
+			//custom code on tile to ask if it aptly got hit, if you want custom geometry
+//
+//What needs to be returned
+//What it hit, game object/tile
+//Normal of hit
+//Can be done quite easily since we know if me moving left or right
+
+//Static manager
+//Calculate if Within bounds of matrix?
+//End and start World pos
+//
+//Calculate offset to then put onto positions, Maybe each position gets its own Calculation
+
+			double RelativeX = 0;
+			double RelativeY = 0;
+
+
+			double gridOffsetx = 0;
+			double gridOffsety = 0;
+
+			int xSteps = 0;
+			int ySteps = 0;
+
+			int stepX = 0;
+			int stepY = 0;
+
+
+			double Offsetuntouchx = (origin.x - Math.Round(origin.x));
+			double Offsetuntouchy = (origin.y - Math.Round(origin.y));
+
+			double offsetX = 0; //Might be conditional
```

Comment:
Codacy found an issue: [Remove the unused local variable 'offsetX'.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6319358)